src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs

```diff
@@ -28,6 +28,346 @@ public static Vector<T> InitVector<T>(Func<int, T> f)
+                }
+            }
+
+             var result = new ushort[count];
+            Array.Copy(srcMask, 0, result, 0, count);
+
+            if (first >= 0)
+            {
+                result[first] = 1;
+            }
+            return result;
+        }
+
+        public static uint[] CreateMaskForFirstActiveElement(uint[] mask, uint[] srcMask)
+        {
+            var count = Vector<uint>.Count;
+
+            var first = -1;
+
+            for (var i = 0; i < count; i++)
+            {
+                if (mask[i] > 0 && first == -1)
+                {
+                    first = i;
+                    break;
+                }
+            }
+
+            var result = new uint[count];
+            Array.Copy(srcMask, 0, result, 0, count);
+
+            if (first >= 0)
+            {
+                result[first] = 1;
+            }
+            return result;
+        }
+
+        public static ulong[] CreateMaskForFirstActiveElement(ulong[] mask, ulong[] srcMask)
+        {
+            var count = Vector<ulong>.Count;
+
+            var result = new ulong[count];
+            Array.Copy(srcMask, 0, result, 0, count);
+
+            var first = -1;
+
+            for (var i = 0; i < count; i++)
+            {
+                if (mask[i] > 0 && first == -1)
+                {
+                    first = i;
+                    break;
+                }
+            }
+
+            if (first >= 0)
+            {
+                result[first] = 1;
+            }
+            return result;
+        }
+
+        public static int LastActiveElement(byte[] v)
+        {
+            var count = Vector<byte>.Count;
+
+            for (var i = count - 1; i >= 0; i--)
+            {
+                if (v[i] > 0)
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        public static int LastActiveElement(ushort[] v)
+        {
+            var count = Vector<ushort>.Count;
+
+            for (var i = count - 1; i >= 0; i--)
+            {
+                if (v[i] > 0)
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        public static int LastActiveElement(uint[] v)
+        {
+            var count = Vector<uint>.Count;
+
+            for (var i = count - 1; i >= 0; i--)
+            {
+                if (v[i] > 0)
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        public static int LastActiveElement(ulong[] v)
+        {
+            var count = Vector<ulong>.Count;
+
+            for (var i = count - 1; i >= 0; i--)
+            {
+                if (v[i] > 0)
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        public static byte[] CreateMaskForNextActiveElement(byte[] mask, byte[] srcMask)
```

Comment:
This might be easier to undersand:

```c#
public static ushort[] CreateMaskForNextActiveElement(ushort[] mask, ushort[] srcMask)
{    
    int activeElement = 0;
    int i;
    // Find the active element in srcMask
    for (i = 0; i < count; i++)
    {
        if (srcMask[i] == 0)
        {
            activeElement = i + 1;
            break;
        }
    }
    
    // Find the next active element in mask
    var result = new byte[Vector<ushort>.Count];
    for (i = activeElement; i < count; i++)
    {
        if (mask[i] == 1)
        {
            result[i] = 1;
            break;
        }
    }
    
    return result;
}
```

Reply:
I was pretty much verbatim copying the logic from the arm docs.

Reply:
> This might be easier to undersand:

Looks good to me (ditto for the simpler `CreateMaskForFirstActiveElement`)

Reply:
The suggestion for `CreateMaskForFirstActiveElement` mutates the `srcMask` which we do not want to do. I'll cleanup these methods though.

Reply:
> The suggestion for CreateMaskForFirstActiveElement mutates the srcMask which we do not want to do

does it matter for our testing?

Reply:
> I was pretty much verbatim copying the logic from the arm docs.

👍. The arm architecture manuals describe a behavior using pseudo-code already and so matching that as closely as possible in the test validation is general goodness and far more beneficial in practice than trying to make rewrite the validation APIs to some form of more aesthetically pleasing code.

Reply:
We explicitly do the same for xarch intrinsics for the same reason, a lot of the logic for these intrinsics is quite complex and we don't want to introduce risk by trying to do any cleanup as compared to what is actually documented to occur (as that can lead to subtle bugs that are difficult to track down).

Reply:
It does. I expect these helpers to not mutate any inputs.

Reply:
> The arm architecture manuals describe a behavior using pseudo-code already and so matching that as closely as possible in the test validation is general goodness and far more beneficial in practice than trying to make rewrite the validation APIs to some form of more aesthetically pleasing code.

That should be the intent, but simplifying the code really helps in understanding if we are even validating the right thing. It took me a while to understand the logic here. Plus, when we say "copying the logic", we are implementing it ourselves anyway, and that can have bugs as well.

Reply:
> but simplifying the code really helps in understanding if we are even validating the right thing

There's not really a need to understand what it's doing, the algorithm matches what is spec'd and that is sufficient for knowing its testing the right thing.

> we are implementing it ourselves anyway, and that can have bugs as well.

Yes, and it is easier to validate that it is correct and matches the expected behavior when it is 1-to-1 (or as close to possible) with the architecture manual.

If we have to simplify the code, then every person needs to read the Architecture manual and work through the logic of simplifying it and understanding what the code is meant to be doing.

If we instead simply mirror the algorithm as closely as possible, it becomes far more trivial to do a comparison with the architecture manual and see that it effectively identical and therefore can be presumed correct.

Reply:
Correctness is by far the most important thing here, far more than perf or any other consideration.

If there was a significant need to document a particularly complex algorithm to make it easier for reviewers to understand, that is then a separate consideration and one that can be handled instead with an attached comment (in code) that describes the operation in "simpler" terms, which then gives us something that can be trivially compared with the source of truth while still providing additional context that makes it easier to understand what's happening.